File: 794_valid_tic_tac_toe/valid_tic_tac_toe.go
/*
 * @Date: 2021-12-09 05:48:24
 * @Author: Mengsen Wang
 * @LastEditors: Mengsen Wang
 * @LastEditTime: 2021-12-09 05:57:00
 */

package main

import "strings"

func win(board []string, p byte) bool {
	for i := 0; i < 3; i++ {
		if board[i][0] == p && board[i][1] == p && board[i][2] == p ||
			board[0][i] == p && board[1][i] == p && board[2][i] == p {
			return true
		}
	}
	return board[0][0] == p && board[1][1] == p && board[2][2] == p ||
		board[0][2] == p && board[1][1] == p && board[2][0] == p
}

func validTicTacToe(board []string) bool {
	oCount, xCount := 0, 0
	for _, row := range board {
		oCount += strings.Count(row, "O")
		xCount += strings.Count(row, "X")
	}
	return !(oCount != xCount && oCount != xCount-1 ||
		oCount != xCount && win(board, 'O') ||
		oCount != xCount-1 && win(board, 'X'))
}

func main() {
	assert := func(b bool) {
		if !b {
			panic("Not Passed")
		}
	}
	assert(!validTicTacToe([]string{"O  ", "   ", "   "}))
	assert(!validTicTacToe([]string{"XOX", " X ", "   "}))
	assert(!validTicTacToe([]string{"XXX", "   ", "OOO"}))
	assert(validTicTacToe([]string{"XOX", "O O", "XOX"}))
}

File: cpp/794_valid_tic_tac_toe/valid_tic_tac_toe.cpp
/*
 * @Date: 2021-12-09 05:48:19
 * @Author: Mengsen Wang
 * @LastEditors: Mengsen Wang
 * @LastEditTime: 2021-12-09 06:06:02
 */

#include <cassert>
#include <string>
#include <vector>

using namespace std;

class Solution {
 public:
  bool validTicTacToe(vector<string> board) {
    int xCount = 0, oCount = 0;
    for (string& row : board) {
      for (char c : row) {
        xCount = (c == 'X') ? (xCount + 1) : xCount;
        oCount = (c == 'O') ? (oCount + 1) : oCount;
      }
    }
    if (oCount != xCount && oCount != xCount - 1) return false;
    if (win(board, 'X') && oCount != xCount - 1) return false;
    if (win(board, 'O') && oCount != xCount) return false;
    return true;
  }

  bool win(vector<string>& board, char p) {
    for (int i = 0; i < 3; ++i) {
      if (p == board[i][0] && p == board[i][1] && p == board[i][2]) return true;
      if (p == board[0][i] && p == board[1][i] && p == board[2][i]) return true;
    }
    if (p == board[0][0] && p == board[1][1] && p == board[2][2]) return true;
    if (p == board[0][2] && p == board[1][1] && p == board[2][0]) return true;
    return false;
  }
};

int main() {
  assert(!Solution().validTicTacToe(vector<string>{"O  ", "   ", "   "}));
  assert(!Solution().validTicTacToe(vector<string>{"XOX", " X ", "   "}));
  assert(!Solution().validTicTacToe(vector<string>{"XXX", "   ", "OOO"}));
  assert(Solution().validTicTacToe(vector<string>{"XOX", "O O", "XOX"}));
}
